### PR TITLE
feat(settlement): sequence diagrams for continuous gross settlement

### PR DESCRIPTION
**Pull request by shashi165**
### Summary
 1. added the sequence diagram for CGS handler :  https://github.com/mojaloop/project/issues/2056
 2. added the sequence diagram for rules handler :  https://github.com/mojaloop/project/issues/2056
 3. added README on how to generate ERD diagrams in DBeaver
 4. fixed the bug with `scripts/_build_plantuml_diff.sh`

### CODE REVIEW

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                                else    
+                                    RULES_HANDLER -> RULES_HANDLER: Insert LedgerEntry into Valid batch;    
+                                end
+                            else
+                                RULES_HANDLER -> RULES_HANDLER: Insert LedgerEntry into Valid batch;    
+                            end
+                        end
+                    end
+                    alt validBatch has entries
+                        RULES_HANDLER -> SETTLEMENT_DAO: Insert valid ledger entries
```
Comment (mdebarros) [inline mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:162]: What happens if there is a failure on one of the DB calls?

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                                where `E`.`name` = {ledgerEntry.ledgerEntryTypeId}
+                                  and `P`.`name` in ({ledgerEntry.payerFspId}, {ledgerEntry.payeeFspId})
+                                  and `PC`.`currencyId` = {ledgerEntry.currency};
+                            end hnote
+                            SETTLEMENT_DAO -> DB: Update positions
+                            activate DB
+                            deactivate DB
+                            loop transferParticipant records
+                                hnote over DB #lightyellow
+                                    update **`participantPosition`**
```
Comment (mdebarros) [inline mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:193]: Do we not need any locks here before we update the participantPosition and participantPositionChange?

**Review by mdebarros — CHANGES_REQUESTED** [review]
Hey @shashi165,

It's looking good, but I do have some comments for your consideration.

File: mojaloop-technical-overview/central-settlements/settlement-process/README.md [hunk 3]
```diff
@@ -78,3 +82,12 @@ Used to request drill-down information regarding a settlement, participant and a
...
 - [Sequence Diagram for Query Settlements by Parameters](get-settlements-by-params.md)
+
+### 2.10 Gross Settlement Handler
+
+This handler executes after each transfer is committed and performs the following operations on success:
+ Handle the updating of the POSITION and SETTLEMENT accounts for participants involved in a transfer where there is a settlement model defined as immeditate and gross on the POSITION account to facilitate RTCGS (Rel-Time Continuous Gross Settlement) per transfer.
+
+This is done by consuming events of the notification topic.
+
+[Sequence Diagram for Transfer Settlement Handler](transfer-settlement-handler-consume.md)
```
[the remark this answers is not in the document]
Comment (shashi165) [reply mojaloop-technical-overview/central-settlements/settlement-process/README.md:93]: done..

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                    SCRIPT_ENGINE -> TRANSACTION_DAO: Get transaction object
+                    TRANSACTION_DAO -> TRANSACTION_DAO: decode ILP packet
+                    TRANSACTION_DAO --> SCRIPT_ENGINE: Transaction object
+                    SCRIPT_ENGINE -> SCRIPT_ENGINE: execute script in sandbox
+                    deactivate
+                    SCRIPT_ENGINE --> SCRIPTS_LOADER: Ledger entries
+                    deactivate SCRIPT_ENGINE
+                    SCRIPTS_LOADER -> SCRIPTS_LOADER: Merge results
+                end
+                SCRIPTS_LOADER --> RULES_HANDLER: script results
```
[the remark this answers is not in the document]
Comment (shashi165) [reply mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:122]: At the moment it's always LedgerEntries

[the remark this answers is not in the document]
Comment (mdebarros) [reply mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:122]: Yes, but depending on the rule, there may be "no results". I.e. what happens if there are no "outputs" (i.e. ledger entries) from the script

**Comment by mdebarros** [comment]
> Hey @shashi165,
> 
> It's looking good, but I do have some comments for your consideration.

Hey @shashi165 @vgenev did you address all my questions?

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    end
+    group Process Single Message
+        TOPIC_NOTIFICATION <- RULES_HANDLER: Consume notification event message
+        activate TOPIC_NOTIFICATION
+        deactivate TOPIC_NOTIFICATION
+        group Validate Message
+            RULES_HANDLER <-> RULES_HANDLER: Validate event - Rule: message has payload\n<color #FF0000><b>Error codes:</b> 2001</color>
+        end
+        alt Rule found
+            RULES_HANDLER <-> RULES_HANDLER: check if ScriptMap[type][action][result] has valid script assigned
```
[the remark this answers is not in the document]
Comment (shashi165) [reply mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:99]: done

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml (hunk 1 above)
Comment (shashi165) [reply mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:147]: The whole transaction will rollback

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml (hunk 4 above)
[the remark this answers is not in the document]
Comment (shashi165) [reply mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:122]: If there is no output from the script then we do nothing

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml (hunk 2 above)
Comment (shashi165) [reply mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:178]: Need to discuss

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-gross-settlement-handler.plantuml [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                                                                 on `M`.`settlementGranularityId` = `G`.`settlementGranularityId`
+                                                      inner join `participantCurrency` as `PC1` on `PC1`.`currencyId` = `PC`.`currencyId` and
+                                                                                                   `PC1`.`participantId` = `PC`.`participantId` and
+                                                                                                   `PC1`.`ledgerAccountTypeId` = `M`.`settlementAccountTypeId`
+                                             where (`TP`.`transferId` = {transferId} and (`G`.`name` = 'GROSS'))) AS TR
+                                            ON PP.participantCurrencyId = TR.ParticipantCurrencyId
+                                 inner join `transferStateChange` as `TSC`
+                                            on `TSC`.`transferID` = {transferId} and `TSC`.`transferStateId` = 'SETTLED';
+                    end hnote
+                end
```
[the remark this answers is not in the document]
Comment (shashi165) [reply mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-gross-settlement-handler.plantuml:188]: The whole transaction will rollback

File: mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mdebarros) [reply mojaloop-technical-overview/central-settlements/settlement-process/assets/diagrams/sequence/seq-rules-handler.plantuml:122]: @shashi165 should this not be reflected on the SDs?

**Review by mdebarros — APPROVED** [review]
+1

**Comment by elnyry-sam-k** [comment]
@shashi165 , @vgenev - seeing this issue on CircleCI: https://app.circleci.com/pipelines/github/mojaloop/documentation/1427/workflows/9dedb110-eb64-48fa-9db8-e362bfa43f09/jobs/2694
```
1 plantuml files are out of sync. Please run npm run build:plantuml and try again.
```

Please review

**Comment by elnyry-sam-k** [comment]
@shashi165 thinking of closing this one out.. do you still need it?

**Comment by elnyry-sam-k** [comment]
@mdebarros @shashi165 @vgenev - closing this one out.. please reopen if necessary

**Pull request closed without merging**